amethyst_core/src/transform/systems.rs

```diff
@@ -31,87 +29,106 @@ impl TransformSystem {
         TransformSystem {
             indices: HashMap::default(),
             sorted: Vec::new(),
-            new: Vec::new(),
+            init: BitSet::new(),
+            frame_init: BitSet::new(),
             dead: HashSet::default(),
-            dirty: HashSet::default(),
-            swapped: HashSet::default(),
+            remove_parent: Vec::new(),
+        }
+    }
+
+    fn remove(&mut self, index: usize) {
+        let entity = self.sorted[index];
+        self.sorted.swap_remove(index);
+        if let Some(swapped) = self.sorted.get(index) {
+            self.indices.insert(*swapped, index);
         }
+        self.indices.remove(&entity);
+        self.init.remove(index as u32);
     }
 }
 
 impl<'a> System<'a> for TransformSystem {
     type SystemData = (
         Entities<'a>,
-        ReadStorage<'a, LocalTransform>,
-        ReadStorage<'a, Child>,
-        WriteStorage<'a, Init>,
+        WriteStorage<'a, LocalTransform>,
+        WriteStorage<'a, Parent>,
         WriteStorage<'a, Transform>,
     );
+    fn run(&mut self, (entities, mut locals, mut parents, mut globals): Self::SystemData) {
+        #[cfg(feature = "profiler")]
+        profile_scope!("transform_system");
+
+        let mut parented = Vec::new();
+        let mut noparents = Vec::new();
+        let mut initted = Vec::new();
 
-    fn run(&mut self, (entities, locals, children, mut init, mut globals): Self::SystemData) {
         // Clear dirty flags on `Transform` storage, before updates go in
         (&mut globals).open().1.clear_flags();
 
-        #[cfg(feature = "profiler")]
-        profile_scope!("transform_system");
-        // Checks for entities with a local transform and parent, but no
-        // `Init` component.
-        for (entity, _, child, _) in (&*entities, &locals, &children, !&init).join() {
-            self.indices.insert(entity, self.sorted.len());
-            self.sorted.push((entity, child.parent()));
-            self.new.push(entity);
-        }
-
-        // Deletes entities whose parents aren't alive.
-        for &(entity, _) in &self.sorted {
-            if let Some(child) = children.get(entity) {
-                if !entities.is_alive(child.parent()) || self.dead.contains(&child.parent()) {
-                    let _ = entities.delete(entity);
-                    self.dead.insert(entity);
+        {
+            for (entity, parent) in (&*entities, parents.open().1).join() {
```

Comment:
We need `retain`!

Reply:
But isn't this too heavy given that it would be very rare?

Reply:
It is only running on modified parents, so I don't think it should be that heavy.

Reply:
I'm not sure yet if `FlaggedStorage` is user friendly enough tbh.

Reply:
Yeah, the issue is you'd have to expose the underlying UnprotectedStorage methods onto Storage if you wanted it to be more straightforward.